# Чтение данных

In [2]:
import pandas as pd

In [8]:
df = pd.read_csv('content/dataset_1_lesson.csv')

In [11]:
df.head()


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [10]:
df


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [12]:
df.isnull().sum()


Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [33]:
for col in df.columns:
    isnull_count = df[col].isnull().sum()
    if isnull_count:
       print(f"Столбец '{col}': нулевых значений {isnull_count}")

Столбец 'LotFrontage': нулевых значений 259
Столбец 'Alley': нулевых значений 1369
Столбец 'MasVnrType': нулевых значений 872
Столбец 'MasVnrArea': нулевых значений 8
Столбец 'BsmtQual': нулевых значений 37
Столбец 'BsmtCond': нулевых значений 37
Столбец 'BsmtExposure': нулевых значений 38
Столбец 'BsmtFinType1': нулевых значений 37
Столбец 'BsmtFinType2': нулевых значений 38
Столбец 'Electrical': нулевых значений 1
Столбец 'FireplaceQu': нулевых значений 690
Столбец 'GarageType': нулевых значений 81
Столбец 'GarageYrBlt': нулевых значений 81
Столбец 'GarageFinish': нулевых значений 81
Столбец 'GarageQual': нулевых значений 81
Столбец 'GarageCond': нулевых значений 81
Столбец 'PoolQC': нулевых значений 1453
Столбец 'Fence': нулевых значений 1179
Столбец 'MiscFeature': нулевых значений 1406


# Пройдемcя по нулевым значеняим:
LotFrontage - Линейные футы улицы, примыкающей к участку - дом может не примыкать к улице. Проверяем эту гипотезу ниже сравнивая с данными LotConfig
'Alley' - Тип доступа к аллее. Нулевое поле просто обозначает что доступа к алее нет.
MasVnrType - тип каменной кладки. Ее может не быть, поэтому нулевые значения оправданы.
MasVnrArea - площадь каменной кладки. Если кладки нет, то и площади нет. (ниже проверяем нет ли исключений что кладка есть, а площадь нулевая)
BsmtQual,BsmtCond, BsmtExposure, BsmtExposure, BsmtFinType1,BsmtFinType2   данные подвала. Нет подвала нет и этих данных(ниже проверяем нет ли исключений т.е отсутствует что-то одно из данных (они или есть оба или обоих нет)
FireplaceQu - качество камина. Нет камина нет качества. ниже проеряем коректность связи камина и качества
Electrical - электрика. Возможно дом без электричества.
GarageType, GarageYrBlt, GarageFinish, GarageQual, GarageCond - данные по граражу. Нет гаража нет данных
PoolQC - качество бассейна. Нет бассейна нет качества (ниже проверяем соответвие бассейнов)
Fence - наличие забора (его может и не быть)
MiscFeature - остальные данные (может и не быть)

# Вывод пропущенных данные - нет.

In [38]:
# смотрим что у нас етсь в LotConfig
df['LotConfig'].unique()


array(['Inside', 'FR2', 'Corner', 'CulDSac', 'FR3'], dtype=object)

In [40]:
# Посмотрим, где пропуски
df.groupby('LotConfig')['LotFrontage'].apply(lambda x: x.isnull().sum())


LotConfig
Corner      62
CulDSac     49
FR2         14
FR3          0
Inside     134
Name: LotFrontage, dtype: int64

Corner: 62 . Это угловые участки. Отсуствие границы с улицей вполне может быть.
CulDSac: 49 Это тупики, поэтому аналогично.
FR2: 14 Примыкание с 2 сторон. Тут не знаю.... спорно, но навреное допустимо.
FR3: 0
Inside: 134 Внутренние участки. Скорее всего тут как раз пропуски - пропущенные данные. Стоит рассмотреть необходимость заполнение их данными (например медианными)


In [ ]:
# Заполнение медианными данными (пока закоменчено)

# df['LotFrontage'] = df.groupby('LotConfig')['LotFrontage'].transform(
#     lambda x: x.fillna(x.median())
# )

In [37]:
mismatch = df[(df['PoolArea'] == 0) & (df['PoolQC'].notnull())]

if not mismatch.empty:
    print("Связь между 'PoolArea' и 'PoolQC' не корректна.")
    print(mismatch)
else:
    print("Связь между 'PoolArea' и 'PoolQC' корректна.")

Связь между 'PoolArea' и 'PoolQC' корректна.


In [36]:
mismatch = df[(df['GarageType'] == 0) & (df['GarageYrBlt'].notnull()) & (df['GarageFinish'].notnull()) & (df['GarageCars'].notnull()) & (df['GarageArea'].notnull()) & (df['GarageQual'].notnull()) & (df['GarageCond'].notnull())]

if not mismatch.empty:
    print("Связь между 'GarageType','GarageYrBlt','GarageFinish','GarageCars','GarageArea','GarageQual','GarageCond' не корректна.")
    print(mismatch)
else:
    print("Связь между 'GarageType','GarageYrBlt','GarageFinish','GarageCars','GarageArea','GarageQual','GarageCond' корректна.")

Связь между 'GarageType','GarageYrBlt','GarageFinish','GarageCars','GarageArea','GarageQual','GarageCond' корректна.


In [34]:
mismatch = df[(df['Fireplaces'] == 0) & (df['FireplaceQu'].notnull())]

if not mismatch.empty:
    print("Связь между 'Fireplaces' и 'FireplaceQu' не корректна.")
    print(mismatch)
else:
    print("Связь между 'Fireplaces' и 'FireplaceQu' корректна.")

Связь между 'Fireplaces' и 'FireplaceQu' корректна.


In [35]:
# проверяем есть ли исключения по подвалу)
mismatch = df[(df['BsmtQual'] == 0) & (df['BsmtCond'].notnull()) & (df['BsmtExposure'].notnull()) & (df['BsmtFinType1'].notnull()) & (df['BsmtFinType2'].notnull())]

if not mismatch.empty:
    print("Связь между 'BsmtQual' и 'BsmtCond', 'BsmtExposure', 'BsmtFinType1','BsmtFinType2' не корректна.")
    print(mismatch)
else:
    print("Связь между 'BsmtQual' и 'BsmtCond', 'BsmtExposure', 'BsmtFinType1','BsmtFinType2' корректна.")
    


Связь между 'BsmtQual' и 'BsmtCond', 'BsmtExposure', 'BsmtFinType1','BsmtFinType2' корректна.


In [28]:
# проверяем есть ли пропущенные данные в каменнйо кладке и ее площади 
#(т.е проверка если кладка есть,а  площадь 0)
for index, row in df.iterrows():
    if row['MasVnrArea'] == '0'  and row['MasVnrType'] != 'None':
        print(f"Пропущены данные MasVnrArea = {row['MasVnrArea']}, MasVnrType = {row['MasVnrType']} т.к кладка есть, а площади нет")
   

In [15]:
for col in df.columns:
    unique_count = df[col].nunique()
    print(f"Столбец '{col}': уникальных значений {unique_count}")
    

Столбец 'Id': уникальных значений 1460
Столбец 'MSSubClass': уникальных значений 15
Столбец 'MSZoning': уникальных значений 5
Столбец 'LotFrontage': уникальных значений 110
Столбец 'LotArea': уникальных значений 1073
Столбец 'Street': уникальных значений 2
Столбец 'Alley': уникальных значений 2
Столбец 'LotShape': уникальных значений 4
Столбец 'LandContour': уникальных значений 4
Столбец 'Utilities': уникальных значений 2
Столбец 'LotConfig': уникальных значений 5
Столбец 'LandSlope': уникальных значений 3
Столбец 'Neighborhood': уникальных значений 25
Столбец 'Condition1': уникальных значений 9
Столбец 'Condition2': уникальных значений 8
Столбец 'BldgType': уникальных значений 5
Столбец 'HouseStyle': уникальных значений 8
Столбец 'OverallQual': уникальных значений 10
Столбец 'OverallCond': уникальных значений 9
Столбец 'YearBuilt': уникальных значений 112
Столбец 'YearRemodAdd': уникальных значений 61
Столбец 'RoofStyle': уникальных значений 6
Столбец 'RoofMatl': уникальных значений 8

Столбцов с не уникальным данными нет.